src/main/java/com/puppycrawl/tools/checkstyle/utils/XpathUtil.java

```diff
@@ -0,0 +1,117 @@
+     * There are three types of AST tokens:
+     * 1. Tokens for which the values are equal to the name of the token. Or in other words,
+     * nodes for which the following expression is always true:
+     * <pre>
+     *     detailAst.getText().equals(TokenUtil.getTokenName(detailAst.getType()))
+     * </pre>
+     * For example:
+     * <pre>
+     *     //MODIFIERS[@text='MODIFIERS']
+     *     //OBJBLOCK[@text='OBJBLOCK']
+     * </pre>
+     * These tokens do not match required criteria because their values do not carry any
+     * information, they do not affect the xpath requests and do not help to get more accurate
+     * results. The values of these nodes are useless. No matter which class you analyze, these
+     * values are always the same.
+     * In addition, they make xpath queries more complex, less readable and less obvious.
+     * 2. Tokens for which the values differ from token names, but values are always constant.
+     * For example:
+     * <pre>
+     *     //LITERAL_VOID[@text='void']
+     *     //RCURLY[@text='}']
+     * </pre>
+     * These tokens are not used for the same reasons as were described in the previous part.
+     * 3. Tokens for which values are not constant. The values of these nodes are closely related
+     * to a concrete class, method, variable and so on.
+     * For example:
+     * <pre>
+     *     String greeting = "HelloWorld";
+     *     //STRING_LITERAL[@text='HelloWorld']
+     * </pre>
+     * <pre>
+     *     int year = 2017;
+     *     //NUM_INT[@text=2017]
+     * </pre>
+     * <pre>
+     *     int age = 23;
+     *     //NUM_INT[@text=23]
+     * </pre>
+     * As you can see same {@code NUM_INT} token type can have different values, depending on
+     * context.
+     * <pre>
+     *     public class MyClass {}
+     *     //IDENT[@text='MyClass']
+     * </pre>
+     * Only these tokens support value attribute because they make our xpath queries more accurate.
+     * These token types are listed below.
+     * */
+    private static final List<Integer> TOKEN_TYPES_WITH_VALUE_ATTRIBUTE = Arrays.asList(
+            TokenTypes.IDENT, TokenTypes.STRING_LITERAL, TokenTypes.CHAR_LITERAL,
+            TokenTypes.NUM_LONG, TokenTypes.NUM_INT, TokenTypes.NUM_DOUBLE, TokenTypes.NUM_FLOAT);
+
+    /** Stop instances being created. **/
+    private XpathUtil() {
+    }
+
+    /**
+     * Checks, if specified node can have {@code @text} attribute.
+     *
+     * @param ast {@code DetailAst} element
+     * @return true if element supports {@code @text} attribute, false otherwise
+     */
+    public static boolean supportsValueAttribute(DetailAST ast) {
+        return TOKEN_TYPES_WITH_VALUE_ATTRIBUTE.contains(ast.getType());
+    }
+
+    /**
+     * Returns value of the ast element.
+     *
+     * @param ast {@code DetailAst} element
+     * @return value of the ast element
+     */
+    public static String getValue(DetailAST ast) {
+        String value = ast.getText();
```

Comment:
I am not sure that usage of `value` is good in whole file.
It is better to change it to `text`, we take `text` from Ast, with almost no changes, we should not make new term for this.

if you agree, please make sure there is not word "value" at all in this file.

Reply:
done

Reply:
continued at https://github.com/checkstyle/checkstyle/pull/6357#discussion_r266878094